File: iOS/YouKong/Presentation/Components/PixelCharacter/RiveCharacterView.swift
import SwiftUI
import RiveRuntime

// MARK: - Rive Activity State

/// Rive 动画状态枚举 — 映射后端推断状态到 Rive 状态机输入
enum RiveActivityState: Int, CaseIterable {
    case idle = 0
    case walking = 1
    case eating = 2
    case sleeping = 3
    case working = 4
    case gaming = 5
    case exercising = 6
    case reading = 7
    case drinking = 8
    case cooking = 9
    case driving = 10
    case shopping = 11
    case chatting = 12
    case relaxing = 13
    case running = 14

    var label: String {
        switch self {
        case .idle: return "idle"
        case .walking: return "walking"
        case .eating: return "eating"
        case .sleeping: return "sleeping"
        case .working: return "working"
        case .gaming: return "gaming"
        case .exercising: return "exercising"
        case .reading: return "reading"
        case .drinking: return "drinking"
        case .cooking: return "cooking"
        case .driving: return "driving"
        case .shopping: return "shopping"
        case .chatting: return "chatting"
        case .relaxing: return "relaxing"
        case .running: return "running"
        }
    }

    static func from(_ string: String) -> RiveActivityState {
        return Self.allCases.first { $0.label == string } ?? .idle
    }
}

// MARK: - Rive Character View

/// Rive 动画角色视图 — 从 .riv 文件加载，通过状态机切换动画
struct RiveCharacterView: View {
    let riveState: String
    var gender: String = "male"

    @StateObject private var viewModel = RiveCharacterViewModel()

    var body: some View {
        viewModel.riveView()
            .onAppear {
                viewModel.loadCharacter()
            }
            .onChange(of: riveState) { newState in
                viewModel.setState(newState)
            }
    }
}

// MARK: - ViewModel

@MainActor
class RiveCharacterViewModel: ObservableObject {
    private var riveModel: RiveViewModel?

    func loadCharacter() {
        guard riveModel == nil else { return }
        do {
            riveModel = try RiveViewModel(
                fileName: "character_default",
                stateMachineName: "State Machine 1",
                autoPlay: true
            )
        } catch {
            print("[Rive] Failed to load: \(error)")
        }
    }

    func setState(_ state: String) {
        let activityState = RiveActivityState.from(state)
        try? riveModel?.setInput("stateIndex", value: Double(activityState.rawValue))
    }

    func riveView() -> some View {
        Group {
            if let model = riveModel {
                model.view()
            } else {
                Color.clear
            }
        }
    }
}

File: Android/core/core-ui/src/main/java/com/youkong/core/ui/rive/RiveCharacterView.kt
package com.youkong.core.ui.rive

import android.content.Context
import androidx.compose.runtime.Composable
import androidx.compose.runtime.DisposableEffect
import androidx.compose.runtime.remember
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.viewinterop.AndroidView
import app.rive.runtime.kotlin.RiveAnimationView
import app.rive.runtime.kotlin.core.Rive
import com.youkong.core.ui.R

/**
 * Rive 活动状态枚举 — 映射后端推断状态到 Rive 状态机输入
 *
 * 状态机设计：
 * - 输入名: "stateIndex" (Number)
 * - 每个数字对应一个动画状态
 */
enum class RiveActivityState(val index: Int, val label: String) {
    IDLE(0, "idle"),
    WALKING(1, "walking"),
    EATING(2, "eating"),
    SLEEPING(3, "sleeping"),
    WORKING(4, "working"),
    GAMING(5, "gaming"),
    EXERCISING(6, "exercising"),
    READING(7, "reading"),
    DRINKING(8, "drinking"),
    COOKING(9, "cooking"),
    DRIVING(10, "driving"),
    SHOPPING(11, "shopping"),
    CHATTING(12, "chatting"),
    RELAXING(13, "relaxing"),
    RUNNING(14, "running");

    companion object {
        fun fromString(state: String): RiveActivityState {
            return entries.find { it.label == state } ?: IDLE
        }
    }
}

/**
 * Rive 角色动画视图
 *
 * @param riveState 后端传来的状态名 (如 "eating", "walking")
 * @param gender 性别 "male" / "female"，用于选择不同 .riv 文件
 * @param modifier Compose modifier
 */
@Composable
fun RiveCharacterView(
    riveState: String,
    gender: String = "male",
    modifier: Modifier = Modifier,
) {
    val context = LocalContext.current
    val activityState = remember(riveState) { RiveActivityState.fromString(riveState) }

    // Initialize Rive runtime
    remember {
        try {
            Rive.init(context)
        } catch (_: Exception) {
            // Already initialized
        }
        true
    }

    val riveView = remember {
        RiveAnimationView(context).apply {
            try {
                setRiveResource(
                    resId = R.raw.character_default,
                    stateMachineName = "State Machine 1",
                    autoplay = true,
                )
            } catch (e: Exception) {
                android.util.Log.e("RiveCharacter", "Failed to load .riv: ${e.message}")
            }
        }
    }

    // Update state when riveState changes
    DisposableEffect(activityState) {
        try {
            riveView.setNumberState("State Machine 1", "stateIndex", activityState.index.toFloat())
        } catch (e: Exception) {
            android.util.Log.d("RiveCharacter", "State machine input not found: ${e.message}")
        }
        onDispose { }
    }

    DisposableEffect(Unit) {
        onDispose {
            // Rive view cleanup handled by Android lifecycle
        }
    }

    AndroidView(
        factory = { riveView },
        modifier = modifier,
    )
}
